src/graphics.c

```diff
@@ -516,14 +531,29 @@ GdipRestoreGraphics (GpGraphics *graphics, GraphicsState state)
 
 	GdipSetRenderingOrigin (graphics, pos_state->org_x, pos_state->org_y);
 
-	if (graphics->clip)
+	if (graphics->clip) {
+		if (graphics->overall_clip == graphics->clip)
+			graphics->overall_clip = NULL; // Don't double-free!
 		GdipDeleteRegion (graphics->clip);
+	}
 	status = GdipCloneRegion (pos_state->clip, &graphics->clip);
 	if (status != Ok)
 		return status;
 
+	if (graphics->previous_clip) {
+		GdipDeleteRegion (graphics->previous_clip);
+		graphics->previous_clip = NULL;
+	}
+	if (pos_state->previous_clip) {
+		status = GdipCloneRegion (pos_state->previous_clip, &graphics->previous_clip);
+		if (status != Ok)
+			return status;
+	}
+
 	gdip_cairo_matrix_copy (graphics->clip_matrix, &pos_state->clip_matrix);
 
+	gdip_calculate_overall_clipping (graphics);
```

Comment:
`graphics->overall_clip` is leaked at this point. Couple of lines above you set `graphics->previous_clip` to `NULL`. This causes `gdip_calculate_overall_clipping` to enter this code path:

```
	if (!graphics->previous_clip) {
		graphics->overall_clip = graphics->clip;
	}
```

and never free the original `overall_clip`.